subprojects/gdk_rust/gdk_electrum/src/spv.rs

```diff
@@ -0,0 +1,338 @@
+                    },
+                ) if *r_work > p_work => curr_result,
+
+                // Otherwise, stick with what we have
+                (r, _) => r,
+            };
+
+            // XXX break early if the result is severe enough already?
+        }
+
+        // Give some grace for minor digressions from the longest chain
+        // XXX determine exact logic
+        match final_result {
+            CrossValidationResult::Lagging {
+                longest_height,
+                ..
+            } if longest_height - chain.height() == 1 => {
+                // Lagging behind the longest chain by 1 block
+                final_result = CrossValidationResult::Valid
+            }
+            CrossValidationResult::MinorityFork {
+                common_ancestor,
+                longest_height,
+                ..
+            } if common_ancestor == longest_height - 1 => {
+                // A shallow fork with a depth of 1 (two blocks competing at the tip)
+                // XXX this should never actually happen, because the two blocks at the tip should have the same difficulty
+                final_result = CrossValidationResult::Valid
+            }
+            _ => (),
+        };
+
+        Ok(final_result)
+    }
+}
+
+pub fn spv_cross_validate(
+    chain: &HeadersChain,
+    local_tip_hash: &BlockHash,
+    server_url: &ElectrumUrl,
+) -> Result<CrossValidationResult, CrossValidationError> {
+    let client = server_url.build_client()?;
+    let remote_tip = client.block_headers_subscribe()?;
+    let remote_tip_hash = remote_tip.header.block_hash();
+    let remote_tip_height = remote_tip.height as u32;
+
+    // Both point to the same tip
+    if remote_tip_height == chain.height() && remote_tip_hash == *local_tip_hash {
+        return Ok(CrossValidationResult::Valid);
+    }
+
+    // The remote tip is lagging behind the local tip and can be fast-forwarded to it
+    if chain.height() > remote_tip_height {
+        let local_header = chain.get(remote_tip_height)?;
+        if local_header.block_hash() == remote_tip_hash {
+            return Ok(CrossValidationResult::Valid);
+        }
+    }
+
+    // The local tip is lagging behind the remote tip and can be fast-forwarded to it
+    if chain.height() < remote_tip_height {
+        let remote_header = client.block_header(chain.height() as usize)?;
+        if remote_header.block_hash() == *local_tip_hash {
+            let fork = get_fork_branch(chain, &client, remote_tip_height, Some(chain.height()))?;
+
+            return Ok(CrossValidationResult::Lagging {
+                longest_height: fork.tip_height,
+                work_diff: fork.total_fork_work,
+            });
+        }
+    }
+
+    let fork = get_fork_branch(chain, &client, remote_tip_height, None)?;
+
+    let our_work: Uint256 = (fork.common_ancestor + 1..=chain.height())
+        .fold(Uint256::zero(), |total, height| total + chain.get(height).unwrap().work());
```

Comment:
I think it's better to remove this unwrap, either by removing the closure or using try_fold?

Reply:
It is a logical error for any of these blocks not to exists, since it iterates up to a upper bound reported by HeadersChain as its tip height. If they are somehow missing, it would indicate that the state got irrevocably corrupted, in which case I think that panicking and crashing the whole thing down is a reasonable thing to do.

Reply:
why not unwrapping at line 158 then?
I have no strong position but I am not convinced either, some file system error for example could be treated by catching the error and deleting the file

Reply:
> why not unwrapping at line 158 then?

I would actually change line 158 to `unwrap()` too.

Another reason for this is that I defined `CrossValidationError` to mean an error that indicates something is wrong with the secondary server and not with the primary server/chain, so its kinda strange that a missing expected local header would result in a variant of `CrossValidationError`. Though the definition could of course be changed if deemed necessary.

> some file system error for example could be treated by catching the error and deleting the file

Wouldn't you also need to restore the file's contents after deleting it?

But in any case, until we have the error recovery logic implemented, if we somehow end up with corrupted database state and the choice is between letting the program continue doing "business as usual" without a specialized recovery procedure or crashing it, I would opt to crash it.

Reply:
ok, fine